File: stat_model/LinReg.m
% Author: Ricardo Baptista and Matthias Poloczek
% Date:   June 2018
%
% See LICENSE.md for copyright information
%

classdef LinReg

	properties

		estimator	% type of estimator: mle, bayes, horseshoe, sparseHS
		order 		% maximum order of monomials included in model
		nVars		% number of input variables
		nCoeffs 	% number of parameters to estimate

		xTrain		% input training data - predictors/covariates
		yTrain 		% input training data - output

		xInf  		% X data with Inf output value - separate from training
		yInf 		% Y data with Inf output value - separate from training

		alpha_mle 	% MLE alpha vector
		alpha_hs    % horseshoe alpha vector
		alphaMean   % bayes posterior alpha mean
		alphaCov    % bayes posterior alpha covariance
		a 	 		% bayes posterior a hyper-parameter
		b  			% bayes posterior b hyper-parameter

	end

	methods
		function LR = LinReg(nVars, order, estimator, varargin)

			p = ImprovedInputParser;
			addRequired(p,'nVars');
			addRequired(p,'order');
			addRequired(p,'estimator');
			parse(p,nVars,order,estimator,varargin{:});
			LR = passMatchedArgsToProperties(p, LR);

		end %endFunction
		% --------------------------------------------------------------
		% --------------------------------------------------------------
		function LR = setupData(LR)

			% limit data to unique points
			[X, x_idx] = unique(LR.xTrain,'rows');
			y = LR.yTrain(x_idx);

			% set upper threshold
			infT = 1e6;

			% separate samples based on Inf output
			y_Infidx  = find(abs(y) > infT);
			y_nInfidx = setdiff(1:length(y), y_Infidx);

			% save samples in LR
			LR.xInf = X(y_Infidx,:);
			LR.yInf = y(y_Infidx);

			LR.xTrain = X(y_nInfidx,:);
			LR.yTrain = y(y_nInfidx);

		end %endFunction
		% --------------------------------------------------------------
		% --------------------------------------------------------------
		function LR = train(LR, inputs)

			% Set nSamps (Gibbs iterations to run)
			inputs.nSamps = 1e3;

			% set data
			LR.xTrain = inputs.x_vals;
			LR.yTrain = inputs.y_vals;

			% setup data for training
			LR = LR.setupData();

			% compute all covariates based on order
			LR.xTrain = order_effects(LR.xTrain, LR.order);
			if strcmp(LR.estimator, 'bayes') || strcmp(LR.estimator, 'mle')
				LR.xTrain = [ones(size(LR.xTrain,1),1), LR.xTrain];
			end

			% find the total number of coefficients
			LR.nCoeffs = size(LR.xTrain,2);

			% run training for different estimators
			if strcmp(LR.estimator, 'bayes')
				LR = LR.bayesTrain(inputs.aPr, inputs.bPr);
			elseif strcmp(LR.estimator, 'mle')
				LR = LR.mleTrain();
			elseif (strcmp(LR.estimator, 'horseshoe') || strcmp(LR.estimator, 'sparseHS'))
				LR = LR.horseshoeTrain(inputs.nSamps);
			else
				error('Estimator is not implemented!')
			end

		end %endFunction
		% --------------------------------------------------------------
		% --------------------------------------------------------------
		function LR = bayesTrain(LR, aPr, bPr)

			nCoeffs_ = LR.nCoeffs;

			% set alpha prior parameters
			aMeanPr = (pi/2)/(nCoeffs_)*ones(nCoeffs_,1);
			aCovPr  = eye(nCoeffs_);
			aInvPr  = inv(aCovPr);

			% extract data
			X = LR.xTrain;
			y = LR.yTrain;
			n = size(X,1);

			% compute alpha posterior parameters
			alphaPrec_ = aInvPr + X'*X;
			alphaCov_  = aCovPr - aCovPr*X'*((eye(n) + X*aCovPr*X')\(X*aCovPr)); % Sherman-Morrison formula
			alphaMean_ = alphaCov_*(aInvPr*aMeanPr + X'*y);

			% symmetrizing covariance
			alphaCov_  = 0.5*(alphaCov_ + alphaCov_');

			% compute hyper-parameter posterior parameters
			aPost = aPr + n/2;
			bPost = bPr + 0.5*(aMeanPr'*aInvPr*aMeanPr + y'*y - alphaMean_'*alphaPrec_*alphaMean_);

			% save results in LR
			LR.alphaMean = alphaMean_;
			LR.alphaCov  = alphaCov_;
			LR.a = aPost;
			LR.b = bPost;

		end %endFunction
		% --------------------------------------------------------------
		% --------------------------------------------------------------
		function LR = mleTrain(LR)

			% extract data
			X = LR.xTrain;
			y = LR.yTrain;
			
			% Compute MLE estimator and save in structure
			regParam = 1e-6;
			LR.alpha_mle = (X'*X + regParam*eye(LR.nCoeffs))\(X'*y);
		
		end %endFunction
		% --------------------------------------------------------------
		% --------------------------------------------------------------
		function LR = horseshoeTrain(LR,nSamps)

			% extract data
			nCoeffs_ = LR.nCoeffs;
			X = LR.xTrain;
			y = LR.yTrain;

			[n_samp, orig_coeff] = size(X);

			% check if x_train contains columns with zeros or duplicates
			idx_zero = find(ismember(X',zeros(1,n_samp),'rows'));
			if (~isempty(idx_zero) || size(unique(X','rows'),1) < size(X,2))

				% remove columns of zeroes
				X(:,idx_zero) = [];
				idx_col = idx_zero;

				% find linearly independent columns to keep
				%idx_licol = LR.licols(X);
				%X = X(:,idx_licol);

				% set idx_col
				%idx_nz  = setdiff(1:orig_coeff,idx_zero);      % kept nz columns
				%idx_nli = setdiff(1:length(idx_nz),idx_licol); % non-li columns
				%idx_col = [idx_zero; idx_nz(idx_nli)'];

			else
				idx_col = [];
			end

			% run Gibbs sampler for nSamps steps
			attempt = 1;
			while(attempt)

				try
					[alpha,a0,~,~] = bhs(X, y, nSamps, 0, 1);
				catch
					continue
				end

				% run until alpha matrix does not contain any NaNs
				if ~any(any(isnan(alpha)))
					attempt = 0;
				end

			end
			
			% append zeros back - note alpha(1,:) is linear intercept
			alpha_pad = zeros(nCoeffs_, 1);
			alpha_pad(setdiff(1:nCoeffs_,idx_col)) = alpha(:,end);
			LR.alpha_hs = [a0(end); alpha_pad];

		end %endFunction
		% --------------------------------------------------------------
		% --------------------------------------------------------------
		function col_idx = licols(LR, X)
		% Extract a linearly independent set of columns from matrix X
		%
		%  X: The given input matrix
		%  tol: A rank estimation tolerance. Default=1e-10
		%  idx: The indices (into X) of the extracted columns

			% set tolerance
			tol=1e-10;

			% compute QR decomposition
			[~, R, E] = qr(X,0); 
			if ~isvector(R)
			    diagr = abs(diag(R));
			else
			    diagr = R(1);   
			end

			% rank estimation
			r = find(diagr >= tol*diagr(1), 1, 'last');

			% extract linearly independent columns
			col_idx = sort(E(1:r));

		end %endFunction
		% --------------------------------------------------------------
		% --------------------------------------------------------------
		function alpha = sampleCoeff(LR)

			% Setup objective function
			if strcmp(LR.estimator, 'bayes')
				sigma2 = 1/gamrnd(LR.a, 1/LR.b);
				alpha  = mvnrnd(LR.alphaMean, sigma2*LR.alphaCov)';
			elseif strcmp(LR.estimator, 'mle')
				alpha  = LR.alpha_mle;
			elseif strcmp(LR.estimator, 'horseshoe')
				alpha  = LR.alpha_hs;
			elseif strcmp(LR.estimator, 'sparseHS')
				alpha  = LR.alpha_hs;
				SpThreshold = 0.1;
				alpha(abs(alpha) < SpThreshold) = 0;
			else
				error('Estimator is not implemented!')
			end

		end %endFunction
		% --------------------------------------------------------------
		% --------------------------------------------------------------
		function out = surrogate_model(LR, x, alpha)
		% SURROGATE_MODEL: Function evaluates the linear model
		% Assumption: input x only contains one row

			% generate x_all (all basis vectors) based on model order
			x_all = [1, order_effects(x, LR.order)];

			% check if x maps to an Inf output
			barrier = 0;
			if ismember(LR.xInf, x, 'rows')
				barrier = Inf;
			end

			% compute objective with barrier
			out = x_all*alpha + barrier;

		end %endFunction
		% --------------------------------------------------------------
		% --------------------------------------------------------------
	end %endMethods
end %endClass

File: BOCSpy/LinReg.py
# Author: Ricardo Baptista and Matthias Poloczek
# Date:   June 2018
#
# See LICENSE.md for copyright information
#

import numpy as np
from itertools import combinations
from bhs import bhs

class LinReg:

	def __init__(self, nVars, order):
		self.nVars = nVars
		self.order = order

	# ---------------------------------------------------------
	# ---------------------------------------------------------

	def setupData(self):

		# limit data to unique points
		X, x_idx = np.unique(self.xTrain, axis=0, return_index=True)
		y = self.yTrain[x_idx]

		# set upper threshold
		infT = 1e6

		# separate samples based on Inf output
		y_Infidx  = np.where(np.abs(y) > infT)[0]
		y_nInfidx = np.setdiff1d(np.arange(len(y)), y_Infidx)

		# save samples in two sets of variables
		self.xInf = X[y_Infidx,:]
		self.yInf = y[y_Infidx]

		self.xTrain = X[y_nInfidx,:]
		self.yTrain = y[y_nInfidx]

	# ---------------------------------------------------------
	# ---------------------------------------------------------

	def train(self, inputs):

		# Set nGibbs (Gibbs iterations to run)
		nGibbs = int(1e3)

		# set data
		self.xTrain = inputs['x_vals']
		self.yTrain = inputs['y_vals']

		# setup data for training
		self.setupData()

		# create matrix with all covariates based on order
		self.xTrain = self.order_effects(self.xTrain, self.order)
		(nSamps, nCoeffs) = self.xTrain.shape

		# check if x_train contains columns with zeros or duplicates
		# and find the corresponding indices
		check_zero = np.all(self.xTrain == np.zeros((nSamps,1)), axis=0)
		idx_zero   = np.where(check_zero == True)[0]
		idx_nnzero = np.where(check_zero == False)[0]

		# remove columns of zeros in self.xTrain
		if np.any(check_zero):
			self.xTrain = self.xTrain[:,idx_nnzero]

		# run Gibbs sampler for nGibbs steps
		attempt = 1
		while(attempt):

			# re-run if there is an error during sampling
			try:
				alphaGibbs,a0,_,_,_ = bhs(self.xTrain,self.yTrain,nGibbs,0,1)
			except:
				print('error during Gibbs sampling. Trying again.')
				continue

			# run until alpha matrix does not contain any NaNs
			if not np.isnan(alphaGibbs).any():
				attempt = 0
			
		# append zeros back - note alpha(1,:) is linear intercept
		alpha_pad = np.zeros(nCoeffs)
		alpha_pad[idx_nnzero] = alphaGibbs[:,-1]
		self.alpha = np.append(a0, alpha_pad)

	# ---------------------------------------------------------
	# ---------------------------------------------------------

	def surrogate_model(self, x, alpha):
		# SURROGATE_MODEL: Function evaluates the linear model
		# Assumption: input x only contains one row

		# generate x_all (all basis vectors) based on model order
		x_all = np.append(1, self.order_effects(x, self.order))

		# check if x maps to an Inf output (if so, barrier=Inf)
		barrier = 0.
		if self.xInf.shape[0] != 0:
			if np.equal(x, self.xInf).all(axis=1).any():
				barrier = np.inf

		# compute and return objective with barrier
		out = np.dot(x_all,alpha) + barrier
		
		return out

	# ---------------------------------------------------------
	# ---------------------------------------------------------

	def order_effects(self, x_vals, ord_t):
		# order_effects: Function computes data matrix for all coupling
		# orders to be added into linear regression model.

		# Find number of variables
		n_samp, n_vars = x_vals.shape

		# Generate matrix to store results
		x_allpairs = x_vals

		for ord_i in range(2,ord_t+1):

			# generate all combinations of indices (without diagonals)
			offdProd = np.array(list(combinations(np.arange(n_vars),ord_i)))

			# generate products of input variables
			x_comb = np.zeros((n_samp, offdProd.shape[0], ord_i))
			for j in range(ord_i):
				x_comb[:,:,j] = x_vals[:,offdProd[:,j]]
			x_allpairs = np.append(x_allpairs, np.prod(x_comb,axis=2),axis=1)

		return x_allpairs


# -- END OF FILE --